internal/prometheusbp/spectest/spec.go

```diff
@@ -0,0 +1,203 @@
+	suffixes := []string{"_latency_seconds", "_requests_total", "_active_requests"}
+	names := map[string]struct{}{}
+	for _, suffix := range suffixes {
+		names[prefix+"_"+server+suffix] = struct{}{}
+		names[prefix+"_"+client+suffix] = struct{}{}
+	}
+	return names
+}
+
+// validateName checks the following:
+// 1) the metric name has at least 3 parts separated by "_".
+// The parts are <namespace>_<client/server>_<name>_<suffix>, where suffix is optional.
+// Ref: https://prometheus.io/docs/practices/naming
+// 2) the metric name has the correct prefix.
+// 3) the metric contains either "client" or "server" as the second part.
+func validateName(name, prefix string) error {
+	var batch errorsbp.Batch
+
+	const (
+		separator = "_"
+		partCount = 3
+	)
+	parts := strings.SplitN(name, separator, partCount)
+	if len(parts) < partCount {
+		batch.Add(fmt.Errorf("%w metric %q does not have minimum number (%d) of parts", errLength, name, partCount))
+		return batch.Compile()
+	}
+
+	if got, want := parts[0], prefix; got != want {
+		batch.Add(fmt.Errorf("%w metric %q does not have the expected prefix %q", errPrefix, name, prefix))
+	}
+
+	if parts[1] != client && parts[1] != server {
+		batch.Add(fmt.Errorf("%w metric %q", errClientServer, name))
+	}
+	return batch.Compile()
+}
+
+func validateLabels(name string, gotLabels map[string]string) error {
+	wantLabels := buildLables(name)
+
+	var batch errorsbp.Batch
+	for wantLabel := range wantLabels {
+		if _, ok := gotLabels[wantLabel]; !ok {
+			batch.Add(fmt.Errorf("%w metric %q label %q", errLabelNotFound, name, wantLabel))
+		}
+		delete(wantLabels, wantLabel)
+	}
+
+	if len(wantLabels) > 0 {
+		// todo: err
+	}
+	return batch.Compile()
+}
+
+// buildLables returns a set of expected labels for the metric name provided.
+// prefix is either thrift, http, or grpc.
+// latency_seconds metrics expect the following labels:
+//   - "<prefix>_service"
+//   - "<prefix>_method"
+//   - "<prefix>_success"
+// request_total metrics expect the following labels:
+//   - "<prefix>_service"
+//   - "<prefix>_method"
+//   - "<prefix>_success"
+//   - "<prefix>_exception_type"
+//   - "<prefix>_baseplate_status"
+//   - "<prefix>_baseplate_status_code"
+// active_requests metrics expect the following labels:
+//   - "<prefix>_service"
+//   - "<prefix>_method"
+func buildLables(name string) map[string]struct{} {
+	const (
+		separator = "_"
+		partCount = 3
+	)
+	parts := strings.SplitN(name, separator, partCount)
+	if len(parts) < partCount {
+		// todo: err
+	}
+	prefix := parts[0]
+	suffix := parts[1:]
+
+	labelSuffixes := []string{"service", "method"}
+	successLabelSuffix := "success"
+	switch strings.Join(suffix, separator) {
+	case "_latency_seconds":
+		labelSuffixes = append(labelSuffixes, successLabelSuffix)
+	case "_requests_total":
+		labelSuffixes = append(labelSuffixes, successLabelSuffix, "exception_type", "baseplate_status", "baseplate_status_code")
+	case "_active_requests":
+		// no op
+	default:
```

Comment:
🔕 this is a personal preference/style thing, but I in general prefer to put `default` before all other `case`s in a `switch`. go's `switch` does not require you to exhaust all the possibilities (like rust does), so with or without `default` is a significant difference and having `default` case at the beginning helps highlight that "this `switch` block has a `default` case`.

Reply:
I like the `default` case at the end because then it's a similar logic flow as `if else` blocks, but I don't feel too strongly either way. I'm happy to move the `default` case up if you prefer.